### ISSUE TITLE
⚠️ hcloud controller manager fails when enable_klipper_metal_lb used

### ISSUE THREAD

**Issue by kervel** [kube-hetzner/terraform-hcloud-kube-hetzner#397]
hello,

i tried to create a fresh cluster with (i think) almost default kube.tf.
the nodes seem to get provisioned fine, but the cluster can't run workloads: the controller manager doesn't finish leaving the node unitialized taint on the worker.

i run with three (tiny) control plane nodes (so i can leave os upgrades enabled) and one larger worker node.

this is the output for the controller manager:

```
Flag --allow-untagged-cloud has been deprecated, This flag is deprecated and will be removed in a future release. A cluster-id will be required on cloud instances.
I1104 15:37:57.680191       1 serving.go:348] Generated self-signed cert in-memory
failed to create listener: failed to listen on 0.0.0.0:10258: listen tcp 0.0.0.0:10258: bind: address already in use
Error: failed to create listener: failed to listen on 0.0.0.0:10258: listen tcp 0.0.0.0:10258: bind: address already in use
Usage:
  cloud-controller-manager [flags]

Debugging flags:

      --contention-profiling   Enable lock contention profiling, if profiling is enabled
      --profiling              Enable profiling via web interface host:port/debug/pprof/ (default true)
```

in addition to this, the csi nodes also fail.

```
I1104 15:02:07.178083       1 main.go:113] Version: v2.2.0
W1104 15:02:17.179632       1 connection.go:172] Still connecting to unix:///run/csi/socket
W1104 15:02:27.179284       1 connection.go:172] Still connecting to unix:///run/csi/socket
W1104 15:02:37.179684       1 connection.go:172] Still connecting to unix:///run/csi/socket
W1104 15:02:47.179613       1 connection.go:172] Still connecting to unix:///run/csi/socket
W1104 15:02:57.178812       1 connection.go:172] Still connecting to unix:///run/csi/socket
W1104 15:03:07.178708       1 connection.go:172] Still connecting to unix:///run/csi/socket
W1104 15:03:17.178906       1 connection.go:172] Still connecting to unix:///run/csi/socket
W1104 15:03:27.179627       1 connection.go:172] Still connecting to unix:///run/csi/socket
W1104 15:03:37.178744       1 connection.go:172] Still connecting to unix:///run/csi/socket
W1104 15:03:47.179200       1 connection.go:172] Still connecting to unix:///run/csi/socket
W1104 15:03:57.179595       1 connection.go:172] Still connecting to unix:///run/csi/socket
W1104 15:04:07.179664       1 connection.go:172] Still connecting to unix:///run/csi/socket
```

i did not change a lot on the settings. this is the diff between my kube.tf and the kube.tf.example from

```diff
@@ -103,39 +103,32 @@
 
   agent_nodepools = [
     {
-      name        = "agent-small",
-      server_type = "cpx11",
-      location    = "fsn1",
-      labels      = [],
-      taints      = [],
-      count       = 1
-    },
-    {
       name        = "agent-large",
-      server_type = "cpx21",
+      server_type = "cpx41",
       location    = "nbg1",
       labels      = [],
       # To exclude from Longhorn if that is used for instance. Completely optional.
-      taints = [
-        "server-usage=storage:NoSchedule"
-      ],
+      taints = []
       count       = 1
-    },
-    {
-      name        = "storage",
-      server_type = "cpx21",
-      location    = "fsn1",
-      # Fully optional, just a demo.
-      labels = [
-        "node.kubernetes.io/server-usage=storage"
-      ],
-      count = 1
-      # In the case of using Longhorn, you can use Hetzner volumes instead of using the node's own storage by specifying a value from 10 to 10000 (in GB)
-      # It will create one volume per node in the nodepool, and configure Longhorn to use them.
-      # Something worth noting is that Volume storage is slower than node storage, which is achieved by not mentioning longhorn_volume_size or setting it to 0. 
-      # So for something like DBs, you definitely want node storage, for other things like backups, volume storage is fine, and cheaper.
-      # longhorn_volume_size = 20
     }
+    #{
+    #  name        = "storage",
+    #  server_type = "cpx21",
+    #  location    = "fsn1",
+    #  # Fully optional, just a demo.
+    #  labels = [
+    #    "node.kubernetes.io/server-usage=storage"
+    #  ],
+    #  count = 1
+    #  # In the case of using Longhorn, you can use Hetzner volumes instead of using the node's own storage by specifying a value from 10 to 10000 (in GB)
+    #  # It will create one volume per node in the nodepool, and configure Longhorn to use them.
+    #  # Something worth noting is that Volume storage is slower than node storage, which is achieved by not mentioning longhorn_volume_size or setting it to 0. 
+    #  # So for something like DBs, you definitely want node storage, for other things like backups, volume storage is fine, and cheaper.
+    #  # longhorn_volume_size = 20
+    #}
   ]
 
   # * LB location and type, the latter will depend on how much load you want it to handle, see https://www.hetzner.com/cloud/load-balancer
@@ -163,13 +156,13 @@
   # To use local storage on the nodes, you can enable Longhorn, default is "false".
   # See a full recap on how to configure agent nodepools for longhorn here https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/discussions/373#discussioncomment-3983159
   # ⚠️ Also please note that the latest version of Longhorn is not yet compatible with kubernetes v1.25, so best to set initial_k3s_channel = "v1.24".
-  # enable_longhorn = true
+  enable_longhorn = true
 
   # The file system type for Longhorn, if enabled (ext4 is the default, otherwise you can choose xfs)
   # longhorn_fstype = "xfs"
 
   # how many replica volumes should longhorn create (default is 3)
-  # longhorn_replica_count = 1
+  longhorn_replica_count = 1
 
   # When you enable Longhorn, you can go with the default settings and just modify the above two variables OR you can copy the longhorn_values.yaml.example
   # file to longhorn_value.yaml and put it at the base of your own module, next to your kube.tf, this is Longhorn's own helm values file.
@@ -196,15 +189,15 @@
   # we allow you to add a nginx_ingress_values.yaml file to the root of your module, next to the kube.tf file, it is simply a helm values config file.
   # See the nginx_ingress_values.yaml.example located at the root of this project.
   # After the cluster is deployed, you can always use HelmChartConfig definition to tweak the configuration.
-  # enable_nginx = true
+  enable_nginx = true
 
   # If you want to disable the Traefik ingress controller, to use the Nginx ingress controller for instance, you can can set this to "false". Default is "true".
-  # enable_traefik = false
+  enable_traefik = false
 
   # Use the klipper LB, instead of the default Hetzner one, that has an advantage of dropping the cost of the setup,
   # Automatically "true" in the case of single node cluster.
   # It can work with any ingress controller that you choose to deploy.
-  # enable_klipper_metal_lb = "true"
+  enable_klipper_metal_lb = "true"
 
   # We give you the possibility to use letsencrypt directly with Traefik because it's an easy setup, however it's not optimal,
   # as the free version of Traefik causes a little bit of downtime when when the certificates get renewed. For proper SSL management,
@@ -238,10 +231,10 @@
   # ⚠️ If you are going to use Rancher addons for instance, it's always a good idea to fix the kube version to latest - 0.01,
   # at the time of writing the latest is v1.25, so setting the value below to "v1.24" will insure maximum compatibility with Rancher, Longhorn and so on!
   # As a rule of thumb, avoid using latest, even if it's the default, we advise you to fix it as mentioned above.
-  # initial_k3s_channel = "latest"
+  initial_k3s_channel = "v1.24"
 
   # The cluster name, by default "k3s"
-  # cluster_name = ""
+  cluster_name = "asset360"
 
   # Whether to use the cluster name in the node name, in the form of {cluster_name}-{nodepool_name}, the default is "true".
   # use_cluster_name_in_node_name = false
```

**Comment by mysticaltech** [comment]
@kervel Please, destroy (see readme how to), and run `terraform init -upgrade` and apply again. Also make sure to use the latest terraform cli.

**Comment by mysticaltech** [comment]
@kervel Just saw two errors:

You have a comma missing after taints = []

![ksnip_20221104-180519](https://user-images.githubusercontent.com/518555/200034729-356f1395-0846-4178-b495-0c487db4247b.png)

And you have to use either enable_klipper_metal_lb or enable_nginx. Along with enable_traefik, those 3 are mutually exclusive!

**Comment by kervel** [comment]
i changed the above, which seems to have fixed it, thanks!
about the remarks: the klipper / nginx / traefik:  i tought one needs both an ingress controller and a lb controller ? so  wouldn't nginx + klipper be okay ?

**Comment by kervel** [comment]
ow, posted too soon. i still have the same error (hcloud controller manager crashes due to "address already in use"). however this time, no more taint on the node so i can deploy workloads.

```
Flag --allow-untagged-cloud has been deprecated, This flag is deprecated and will be removed in a future release. A cluster-id will be required on cloud instances.
I1104 20:18:48.016109       1 serving.go:348] Generated self-signed cert in-memory
failed to create listener: failed to listen on 0.0.0.0:10258: listen tcp 0.0.0.0:10258: bind: address already in use
Error: failed to create listener: failed to listen on 0.0.0.0:10258: listen tcp 0.0.0.0:10258: bind: address already in use
```

**Comment by mysticaltech** [comment]
@kervel Yes, you are right, you can use either hetzner for the lbs or klipper. My bad! Please, to facilitate debugging, please do remote all comments from your kube.tf and include attached as a text file without the sensitive values, or just copy-paste it.

**Comment by mysticaltech** [comment]
@kervel You are right, this is happening for me too when enable_klipper_metal_lb (which is quite rare, usually it's just used for single node clusters) but definitely it should work. Will look into it, and keep you posted ASAP.

**Comment by mysticaltech** [comment]
@kervel It turns out this is due to k3s decision to move klipper into a stub ccm, meaning that it cannot be used with the hetzner ccm (hence it fails to bind, as the fake ccm already uses the ccm port 10258 that klipper requires). 

That means that unfortunately that this project will no longer be able to support the use of klipper, and we will have to look at another metal lb. 

So for now, please use the default hetzner lb, by setting `enable_klipper_metal_lb` to false. We will provide an alternative soon.

More info on the matter here https://github.com/k3s-io/k3s/blob/027039530d43bc01c9c17cc0be48db583bff943f/docs/adrs/servicelb-ccm.md.

**Comment by mysticaltech** [comment]
Good news, we may still be able to support klipper! https://github.com/k3s-io/k3s/discussions/6452

**Comment by mysticaltech** [comment]
@kervel This has been fixed in v1.6.4. Enjoy!

**Comment by kervel** [comment]
ow great! thanks!
